internal/throttler/remote/throttler.go

```diff
@@ -0,0 +1,178 @@
+}
+
+type httpCreditManagerProxy struct {
+	hostPort string
+	logger   jaeger.Logger
+}
+
+func newHTTPCreditManagerProxy(hostPort string, logger jaeger.Logger) *httpCreditManagerProxy {
+	return &httpCreditManagerProxy{
+		hostPort: hostPort,
+		logger:   logger,
+	}
+}
+
+func (m *httpCreditManagerProxy) FetchCredits(uuid, serviceName string, operation []string) []creditResponse {
+	params := url.Values{}
+	params.Set("service", serviceName)
+	params.Set("uuid", uuid)
+	for _, op := range operation {
+		params.Add("operation", op)
+	}
+	var resp []creditResponse
+	if err := utils.GetJSON(fmt.Sprintf("http://%s/credits?%s", m.hostPort, params.Encode()), &resp); err != nil {
+		m.logger.Error("Failed to receive credits from agent: " + err.Error())
+	}
+	return resp
+}
+
+// Throttler retrieves credits from agent and uses it to throttle operations.
+type Throttler struct {
+	options
+
+	mux           sync.RWMutex
+	service       string
+	uuid          *atomic.String
+	creditManager *httpCreditManagerProxy
+	credits       map[string]float64 // map of operation->credits
+	close         chan struct{}
+	stopped       sync.WaitGroup
+}
+
+// NewThrottler returns a Throttler that polls agent for credits and uses them to throttle
+// the service.
+func NewThrottler(service string, options ...Option) *Throttler {
+	// TODO add metrics
+	// TODO set a limit on the max number of credits
+	opts := applyOptions(options...)
+	creditManager := newHTTPCreditManagerProxy(opts.hostPort, opts.logger)
+	t := &Throttler{
+		options:       opts,
+		creditManager: creditManager,
+		service:       service,
+		credits:       make(map[string]float64),
+		close:         make(chan struct{}),
+		uuid:          atomic.NewString(""),
+	}
+	t.stopped.Add(1)
+	go t.pollManager()
+	return t
+}
+
+// IsThrottled implements Throttler#IsThrottled.
+func (t *Throttler) IsThrottled(operation string) bool {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	_, ok := t.credits[operation]
+	if !ok {
+		// If it is the first time this operation is being checked, synchronously fetch
+		// the credits.
+		credits := t.fetchCreditsHelper([]string{operation})
+		if len(credits) == 0 {
+			// Failed to receive credits from agent, try again next time
+			return true
+		}
+		t.credits[operation] = credits[0].Credits
+	}
+	return t.isThrottled(operation)
+}
+
+// Close implements Throttler#Close.
+func (t *Throttler) Close() error {
+	close(t.close)
+	t.stopped.Wait()
+	return nil
+}
+
+// SetUUID implements Throttler#SetUUID. It's imperative that the UUID is set before any remote
+// requests are made.
+func (t *Throttler) SetUUID(uuid string) {
+	t.uuid.Store(uuid)
+}
+
+// N.B. This function should be called with the Write Lock
+func (t *Throttler) isThrottled(operation string) bool {
+	credits := t.credits[operation]
+	if credits < minimumCredits {
+		return true
+	}
+	t.credits[operation] = credits - minimumCredits
+	return false
+}
+
+func (t *Throttler) pollManager() {
```

Comment:
Not sure if polling is truly necessary. On the other hand, I believe it should be required to spawn an asynchronous credit request when an operation runs out of credits (i.e. below 1.0). Otherwise, we run the risk of making requests when we don't need credits and not requesting when we do need credits.

Reply:
if we do this on demand, and the demand happens to be very high, there will be a lot of these async requests, each only returning a small fraction of accumulated credits from the agent.

Reply:
also, the agent uses these requests to update the client's updateTime. If we don't ping regularly, we could potentially have the agent give the client way more credits than its due.

Reply:
Ya I was wondering if we could combine the two somehow. I guess let's try this and see if it works well.